compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala

```diff
@@ -99,12 +105,26 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
      * because they are emitted as private by our .scala source files, but
      * they are considered public at use site since their symbols come from
      * Java-emitted .class files.
+     *
+     * Starting with 2.12.0-RC2, we also special case outer fields. This
+     * essentially fix #2382, which is caused by a class having various $outer
+     * pointers in its hierarchy that points to different outer instances.
+     * Without this fix, they all collapse to the same field in the IR. We
+     * cannot fix this for all Scala versions at the moment, because that would
+     * break backwards binary compatibility. We *do* fix it for 2.12.0-RC2
+     * onwards because that also fixes #2625, which surfaced in 2.12 and is
+     * therefore a regression. We can do this because the 2.12 ecosystem is
+     * not binary compatible anyway (because of Scala) so we can break it on
+     * our side at the same time.
      */
-    val idSuffix =
-      if (sym.isPrivate || allRefClasses.contains(sym.owner))
+    val idSuffix: String = {
+      if (applyFixFor2382 && sym.isOuterField)
+        encodeClassFullName(sym.owner)
+      else if (sym.isPrivate || allRefClasses.contains(sym.owner))
         sym.owner.ancestors.count(!_.isTraitOrInterface).toString
```

Comment:
I do not even understand why we need to special case this? Is the Scalac backend special casing this? Why are field overrides with the same name allowed if they are outer fields?